cgmes/cgmes-model/src/main/java/com/powsybl/cgmes/model/CgmesOnDataSource.java

```diff
@@ -109,27 +123,59 @@ public Set<String> names() {
         try {
             // the set of names may be empty if the data source does not contain CGMES data
             Set<String> allNames = dataSource.listNames(REGEX_VALID_NAME);
-            allNames.removeIf(n -> !containsValidNamespace(n));
+            allNames.removeIf(n -> !existsInDatasource(n) || !containsValidNamespace(n));
             return allNames;
         } catch (IOException x) {
-            throw new CgmesModelException(String.format("Listing CGMES names in data source %s", dataSource), x);
+            throw new CgmesModelException(String.format(LISTING_CGMES_NAMES_IN_DATA_SOURCE, dataSource), x);
+        }
+    }
+
+    private boolean existsInDatasource(String fileName) {
+        try {
+            return dataSource.exists(fileName);
+        } catch (IOException e) {
+            return false;
         }
     }
 
     private boolean containsValidNamespace(String name) {
+        String fileExtension = name.substring(name.lastIndexOf('.') + 1);
+        if (fileExtension.equals(CompressionFormat.ZIP.getExtension())) {
+            ZipSecurityHelper.checkIfZipExtractionIsSafe(dataSource, name);
+            try (ZipInputStream is = new ZipInputStream(dataSource.newInputStream(name))) {
+                is.getNextEntry();
+                Set<String> ns = NamespaceReader.namespaces1(is);
+                return ns.contains(RDF_NAMESPACE) && ns.stream().anyMatch(CgmesNamespace::isValid);
+            } catch (XMLStreamException e) {
+                return false;
+            } catch (IOException x) {
+                throw new CgmesModelException(String.format(LISTING_CGMES_NAMES_IN_DATA_SOURCE, dataSource), x);
+            }
+        }
         try (InputStream is = dataSource.newInputStream(name)) {
             Set<String> ns = NamespaceReader.namespaces1(is);
             return ns.contains(RDF_NAMESPACE) && ns.stream().anyMatch(CgmesNamespace::isValid);
         } catch (XMLStreamException e) {
             return false;
         } catch (IOException x) {
-            throw new CgmesModelException(String.format("Listing CGMES names in data source %s", dataSource), x);
+            throw new CgmesModelException(String.format(LISTING_CGMES_NAMES_IN_DATA_SOURCE, dataSource), x);
         }
     }
 
     public Set<String> namespaces() {
         Set<String> ns = new HashSet<>();
         names().forEach(n -> {
+            String fileExtension = n.substring(n.lastIndexOf('.') + 1);
+            if (fileExtension.equals(CompressionFormat.ZIP.getExtension())) {
+                ZipSecurityHelper.checkIfZipExtractionIsSafe(dataSource, n);
+                try (ZipInputStream is = new ZipInputStream(dataSource.newInputStream(n))) {
+                    is.getNextEntry();
```

Comment:
we take the first entry only, shouldn't we check there's only one entry in the zip file? and throw a clear exception otherwise.